ldms/man/ldmsd_sampler_discovery.man

```diff
@@ -0,0 +1,245 @@
+\" Manpage for ldmsd_sampler_discovery
+.TH man 7 "27 March 2024" "v5" "LDMSD Sampler Discovery man page"
+
+.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""/.
+.SH NAME
+ldmsd_sampler_disconvery - Manual for LDMSD Sampler Discovery
+
+.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""/.
+.SH SYNOPSIS
+
+**Sampler side Commands**
+
+.IP \fBadvertise_add
+.RI "name=" NAME " xprt=" XPRT " host=" HOST " port=" PORT
+.RI " [auth=" AUTH_DOMAIN "]"
+
+.IP \fBadvertise_start
+.RI "name=" NAME
+
+.IP \fBadvertise_stop
+.RI "name=" NAME
+
+.IP \fBadvertise_del
+.RI "name=" NAME
+
+.IP \fBadvertise_status
+.RI " [name=" NAME "]"
+
+.PP
+**Aggregator Side Commands**
+
+.IP \fBprdcr_listen-add
+.RI "name=" NAME " regex=" REGEX " " RECONNECT=" INTVL
+.RI " [rail=" SIZE " [credits=" BYTES " ] [rx_rate=" RATE_LIMIT "]"
+
+.IP \fBprdcr_listen_start
+.RI "name=" NAME
+
+.IP \fBprdcr_listen_stop
+.RI "name=" NAME
+
+.IP \fBprdcr_listen_del
+.RI "name=" NAME
+
+.IP \fBprdcr_listen_status
+
+.SH DESCRIPTION
+
+LDMSD Sampler Discovery is a capability that enables LDMSD automatically add
+producers that its hostname matches a given regular expression. The feature
+eliminates the need for manual configuration of sampler hostname in the
+aggregator configuration file.
+
+Admins specify the aggregator hostname and the listening port in sampler
+configuration via the \fBadvertise_add\fR command and start the advertisement
+with the \fBadvertise_start\fR command. The samplers now advertise their
+hostname to the aggregator. On the aggregator, admins specify a regular
+expression to be matched with sampler hostname via the \fBprdcr_listen_add\fR
+command. The \fBprdcr_listen_start\fR command is used to tell the aggregator to
+automatically add producers corresponding to a sampler of which the hostname
+matches the regular expression.
+
+The auto-generated producers is of the ‘generated’ type. The producer name is
+the same as the name given at the \fBadvertise_add\fR line in the sampler
+configuration file. LDMSD automatically starts them; however, admins need to
+stop them manually by using the command \fBprdcr_stop\fR or
+\fBprdcr_stop_regex\fR.  They can be restarted by using the command
+\fBprdcr_start\fR or \fBprdcr_start_regex\fR.
+
+The description for each command and its parameters are as follows.
+
+**Sampler Side Commands**
+
+\fBadvertise_add\fR adds a new advertisement. The parameters are:
+.RS
+.IP \fBname\fR=\fINAME
+String of the advertisement name. The aggregator uses the string as the producer name as well.
+.IP \fBhost\fR=\fIHOST
+Aggregator hostname
+.IP \fBxprt\fR=\fIXPRT
+Transport to connect to the aggregator
+.IP \fBport\fR=\fIPORT
+Listen port of the aggregator
+.IP \fBreconnect\fR=\fIINTERVAL
+Reconnect interval
+.IP \fB[auth\fR=\fIAUTH_DOMAIN\fB]
+The authentication domain to be used to connect to the aggregator
+.RE
+
+\fBadvertise_start\fR starts an advertisement. The parameters are:
```

Comment:
What is the use case for having separate start/stop functions? Why not just start on _add() and stop on _del()? (And maybe then rename: s/add/start, s/del/stop)

Reply:
At first I didn’t implement `start` and `stop` functionality. I didn't even implement `del` as I don't see a use for it. However, when I thought about troubleshooting scenarios, it’s better to have the `stop` command available.

Here are the reasons:

- Separating deletion and stopping from each other allows users to restart the prdcr_listen without typing the whole `add` line again. Use cases of stopping a prdcr_listen happen when things don’t go as planned. For example, the aggregator cannot keep up with updating sets, while more producers are being automatically added to the aggregators. To inspect the situation, I would like to stop adding more producers to the aggregator to prevent adding more load on it. Hence, prdcr_listen_stop would become useful. Whether that will happen, I don’t know. But at some point, we will definitely need to do troubleshooting, and an ability to stop a prdcr_listen could be beneficial.

- Providing prdcr_listen_stop calls for the existence of prdcr_listen_start to restart a stopped prdcr_listen.

Reply:
OK. Since those are pretty uncommon use case, how about instead of having the current add/start/stop/del, have:

  _start(takes all the options that your _add currently takes)
 _pause
 _resume
 _stop

The most common use cases would only do start and maybe stop. But people that need to pause and resume for debugging/whatever can do so.